### ISSUE TITLE
Gen. IV Mystery Gift (again) details and suggestion + FR/LG items quirk

### ISSUE THREAD

**Issue by Ammako** [kwsch/PKHeX#730]
1. Latest build lacks the ability to properly "match" a PCD and a PGT together (this is what I wanted to test last time, before running into that problem with saving the file.)

In Gen. IV the PGT are matched with a PCD, and if the gift that goes with a certain card hasn't been picked up yet, the card will display a message telling you to pick up the gift from the Pokémart (and if it has been picked up, the card will say you have already received this gift.)

Setting three Wondercards in with PKHeX along with three PGT in the same slots, it looks like right now it sets the corresponding card number of PGT1 and PGT3 to 02, while the corresponding card number of PGT2 is set to 00.
In order to have it work the way it's intended to work, the PGT's corresponding card number needs to be set to 00, 01 or 02, depending on which slot their Wondercard is in.
(This is easier to see in PMGE, and the save files I uploaded the other day show clearly how it should normally work.)

Suggestion, could we be able to set PGT directly from PCD without having to separately import a PGT?
For one, this would be a lot more convenient, but at the same time, this could be a way of setting the corresponding card numbers on the PGT slot; when PGT is set from PCD, set that gift slot's corresponding card number to whichever PCD slot was used. If the user specifically sets a PGT without a PCD then I guess corresponding card number could be left as default. (A way to manually specify which card number a PGT corresponds to could be added, in case someone wants to do that I guess.)

2. Items in FR/LG can stack up to 999, however PKHeX only allows you to set up to 95 items per slot, and if you have more than 95 of an item in-game and edit your bags via PKHeX, it will actually "delete" most of them and leave you with only 95.

**Comment by kwsch** [comment]
>Setting three Wondercards in with PKHeX along with three PGT in the same slots, it looks like right now it sets the corresponding card number of PGT1 and PGT3 to 02, while the corresponding card number of PGT2 is set to 00.

PKHeX doesn't modify the "Slot" property, although it may be useful to.

Seems like the smart solution would be when saving the album:
* When setting a PGT (slots 1-8), check all PCDs (slots 9-11) to see if the PGT came from it and if so set the Slot property to match the PCD index.

**Comment by Ammako** [comment]
For the record, I am fairly certain that the default value of 0 refers to the first PCD. Or at least, that's how PMGE shows it :p (I don't think it's a display quirk of PMGE, because even PMGE seems to think that 1 is the first PCD, and thus it does it wrong when setting PGT from PCD.)

I received a bunch of stuff from Shadows of Almia yesterday, the Manaphy Egg, being a PGT with no matching PCD, seems to be given a matching PCD index of 3 (meaning it doesn't point to a valid Wondercard, it doesn't cause any issues.) The other two that are actual Wondercards refer to PCD slot 0 and PCD slot 1.
(This also showed me how it's possible to end up with mismatched PCD and PGT positions; Manaphy Egg does that. Not that it's really important here, anyway.)

**Comment by Ammako** [comment]
(Doesn't seem to work, again Corresponding Card is always set to 0 D: and even PGT with no matching PCD gets it set to 0)

https://puu.sh/ttb8g/7636d3da3d.png
(In this particular screenshot 0 is indeed correct for that WC/Gift, but everything gets set to 0 regardless. Screenshot is mainly just to show which thing I'm talking about exactly.)

Setting PGT from PCD seems to work, thank you for that.
What I had in mind was more along the lines of setting PCD to PCD slot, then having a right-click option to the PCD like "Generate PGT from PCD" which would automatically add a PGT to the next free PGT slot + set corresponding card number to 0/1/2 depending on which slot the PCD was in.
Though if you can get corresponding card numbers to work properly with the way you're doing it right now, then I guess it doesn't matter as much. Your way works too and probably requires less work to implement anyway :p

**Comment by kwsch** [comment]
The form interface has to be generalized enough to support all generations, so having generation-specific menu options isn't really feasible. Users can now drop a PCD to a PGT slot and rearrange them as they see fit, and PKHeX will automatically update the 'associated PCD slot' value of the PGT to the PCD it may be associated with.

**Comment by Ammako** [comment]
Ah yeah, that's more than fair enough.

Do you have the Corresponding Card number/Associated PCD slot setting working on your end though? No matter what I do, that value is always 0 for every PGT slot, whether I set PGT from PCD, or set a standalone PGT (which should be setting it to 3.)
I've tried manually setting a PCD into a PGT slot, and tried dragging from PCD slot to PGT slot, but no dice. Wish I could help figure out why it happens, but I don't understand the code well enough to figure anything out :s

**Comment by kwsch** [comment]
Ha oops, try now.

You had provided a blank save with the 3 crown beasts; the cardMatch returns `int[] { 0, 1, 2, 0, 0, 0, 0, 0 };`

First 3 are lined up with their respective PCD's, and last 5 slots are empty (0).

**Comment by Ammako** [comment]
Oh yeah, much better now. Thanks!

Although I did notice just now, one very specific instance where corresponding card number doesn't seem to get set properly. Not entirely sure how to explain it generally so I'll give as an example what I'm doing in my testing:

I have three Wondercards in my save file, for NZone Pikachu, Gamestp Pichu, and Oblivia Deoxys. I set the PGT from those PCD and save the file, and the associated PCD slot values get properly set. Then I open Mystery Gift editor again, and let's say I delete every PGT and every PCD. If I then right away set a PGT for NZone Pikachu, Gamestp Pichu or Oblivia Deoxys, but without setting a PCD for them, it won't properly set the associated PCD slot value to 3. If I press Save after deleting everything and then re-enter the Mystery Gift edit menu, it will avoid that problem and set the value to 3 as expected.

Basically, if I delete PCDs from the PCD slots, until I press Save the program still thinks that those PCD are present in those slots to some extent and fails to properly set the associated PCD slot value to 3 if I set a PGT that would have matched one of the PCD that were just deleted.

It's a pretty small thing, and I really doubt that people would come across this during normal use, but eh. It might be worth looking into and fixing just to make everything as bullet-proof as possible. :p
